examples/fakerealtime/main.go

```diff
@@ -0,0 +1,225 @@
+// random number (in range (0, N)) to the beginning of the `updates`
+// table primary key. This reduces contention but makes the reader do
+// N queries to see all updates. This is effectively what Friendfeed
+// did, since there was one `updates` table per MySQL shard.
+package main
+
+import (
+	"database/sql"
+	"flag"
+	"fmt"
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/cockroachdb/cockroach/security"
+	"github.com/cockroachdb/cockroach/security/securitytest"
+	"github.com/cockroachdb/cockroach/server"
+	"github.com/cockroachdb/cockroach/util/log"
+	"github.com/montanaflynn/stats"
+)
+
+func createTables(db *sql.DB) error {
+	statements := []string{
+		`create database if not exists fakerealtime`,
+
+		`drop table if exists fakerealtime.updates`,
+		`create table fakerealtime.updates (
+        update_id int,
+        channel string,
+        msg_id int,
+        primary key (update_id, channel)
+    )`,
+
+		`drop table if exists fakerealtime.messages`,
+		`create table fakerealtime.messages (
+        channel string,
+        msg_id int,
+        message string,
+        primary key (channel, msg_id)
+    )`,
+	}
+	for _, stmt := range statements {
+		if _, err := db.Exec(stmt); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+type statistics struct {
+	sync.Mutex
+	writeTimes stats.Float64Data
+}
+
+func (s *statistics) recordWrite(start time.Time) {
+	duration := time.Now().Sub(start)
+	s.Lock()
+	defer s.Unlock()
+	s.writeTimes = append(s.writeTimes, float64(duration.Nanoseconds()))
+}
+
+func (s *statistics) report() {
+	for range time.Tick(time.Second) {
+		s.Lock()
+		writeTimes := s.writeTimes
+		s.writeTimes = nil
+		s.Unlock()
+
+		// The stats functions return an error only when the input is empty.
+		mean, _ := stats.Mean(writeTimes)
+		stddev, _ := stats.StandardDeviation(writeTimes)
+		log.Infof("wrote %d messages, latency mean=%s, stddev=%s",
+			len(writeTimes), time.Duration(mean), time.Duration(stddev))
+	}
+}
+
+type writer struct {
+	db          *sql.DB
+	numChannels int
+	wg          *sync.WaitGroup
+	stats       *statistics
+}
+
+func (w writer) run() {
+	defer w.wg.Done()
+	for {
+		if err := w.writeMessage(); err != nil {
+			log.Errorf("error writing message: %s", err)
+		}
+	}
+}
+
+func (w writer) writeMessage() error {
+	start := time.Now()
+	defer w.stats.recordWrite(start)
+	channel := fmt.Sprintf("room-%d", rand.Int31n(int32(w.numChannels)))
+	message := start.String()
+
+	// TODO(bdarnell): retry only on certain errors.
+	for {
+		txn, err := w.db.Begin()
+		if err != nil {
+			continue
+		}
+
+		// TODO(bdarnell): make this a subquery when subqueries are supported on insert.
+		row := txn.QueryRow(`select max(msg_id) from fakerealtime.messages where channel=$1`, channel)
+		var maxMsgID sql.NullInt64
+		if err := row.Scan(&maxMsgID); err != nil {
+			_ = txn.Rollback()
+			continue
+		}
+		if !maxMsgID.Valid {
+			maxMsgID.Int64 = 0
+		}
+		newMsgID := maxMsgID.Int64 + 1
```

Comment:
For the user experience: it's weird to have messages inserted into the past (and since per-channel contention is so much lower than the global updates table, it doesn't really cost much to be strict here). But it's not really necessary, and depending on what we do for the global queue this could be changed as well.